### DISCUSSION TITLE
Lazy loading

### DISCUSSION THREAD

**Discussion by Josh-Nicholson** [vincjo/datatables#108] [Q&A] [answered]
Hi,

I've been using your project for a while now (its great btw 😊).

I have now encountered the scenario of having to display a table with many thousands of rows.
This of course loads slowly.
So I would like to implement lazy loading.

Is lazy loading possible with this library?

Currently I call off to an api to load all of the results and return it in an array, which I then set with ```handler.setRows(result)```.

I'm assuming that if the current page and row count were tracked by myself rather than the library, I could trigger another call to the api when ever the user changed the page number or the rows per page, then with the result I can use ```setRows``` to update the table reactively.

Although I'm not too sure about how the search functionality would work with lazy loading.

I would appreciate any answers/ideas on the topic.
Cheers

**Comment by riziles** [comment, marked as answer]
@Josh-Nicholson , doesn't the remote API already handle lazy loading pagination and remote search? 
https://vincjo.fr/datatables/remote/examples/user-api

**Reply by Josh-Nicholson** [reply]
Thanks for the reply.

It would appear I didn't read the docs thoroughly😄.

**Reply by vincjo** [reply]
Yes, the documentation for the server-side data processing part is rather incomplete.
It is an extension of the lib that operates independently: all imports are carried out from the namespace `@vincjo/datatables/remote`.

This enables querying an API with each client-side interaction.
The handling of search, pagination, filtering, and sorting is then passed on to the server-side, set in query params.

Features will depend on your API's capabilities.

**Reply by Josh-Nicholson** [reply]
Thanks you for the response. I now have a working lazy loading table. 
Is there a way to show the user a loading animation?
Some users may be using this table over a slow VPN, so I want to show them that something is happening and not that the table is unresponsive. Currently the user changes the page or rows per page and it looks like the table hangs whilst it is waiting for a new set of data.

**Reply by vincjo** [reply]
Good idea, will add a `isLoading` state while data is fetched in a future release.

In the meatime you could add a store to track the loading time, next to your async request/response.
Something like that:

```ts
handler.onChange( (state: State) => {

    isLoading.set(true)

    const response = await fetch(...

    const json = await response.json()
    
    isLoading.set(false)
    
    return json
})
```

#109

**Reply by Josh-Nicholson** [reply]
I've previously implemented custom filters on my other tables (not using the remote library).
I am trying to get one of these custom filters working with my new lazy loading table.
But for some reason the filter change isn't triggering the onChange event.
Can you see any reason why?

```ts

<script lang="ts">
	export let data: PageData;

	const handler = new DataHandler(data.parts, { rowsPerPage: 25, totalRows: data.count });
	const rows = handler.getRows();
	handler.onChange((state: State) => fetchParts(state));

	let selectedActiveFilter: string;

	async function fetchParts(state: State) {
		//call off to api
	}
</script>


//Datatable code
    <p>Active Status:</p>
    <select class="select w-48" name="activeFilter" bind:value={selectedActiveFilter} on:change={() => handler.filter(selectedActiveFilter, 'active')}>
        <option value="">All parts</option>
        <option value="true">Only active parts</option>
        <option value="false">Only inactive parts</option>
    </select>
//end of datatable code

**Reply by Josh-Nicholson** [reply]
Sorry ignore me.

Forgot I had to add the invalidate call.

Here it is working:
```ts
<script lang="ts">
	export let data: PageData;

	const handler = new DataHandler(data.parts, { rowsPerPage: 25, totalRows: data.count });
	const rows = handler.getRows();
	handler.onChange((state: State) => fetchParts(state));

	let selectedActiveFilter: string;

	async function fetchParts(state: State) {
		//call off to api
	}
</script>


//Datatable code
    <p>Active Status:</p>
    <select class="select w-48" name="activeFilter" bind:value={selectedActiveFilter} 
            on:change={() => { handler.filter(selectedActiveFilter, 'active'); handler.invalidate(); }}>
        <option value="">All parts</option>
        <option value="true">Only active parts</option>
        <option value="false">Only inactive parts</option>
    </select>
//end of datatable code
```
